Client/game_sa/CVehicleSA.cpp

```diff
@@ -165,22 +165,34 @@ void CVehicleSA::Init()
     m_RGBColors[3] = CVehicleColor::GetRGBFromPaletteIndex(GetVehicleInterface()->m_colour4);
     SetColor(m_RGBColors[0], m_RGBColors[1], m_RGBColors[2], m_RGBColors[3], 0);
 
-    // Initialize doors depending on the vtable.
-    DWORD dwOffset;
-    DWORD dwFunc = ((CVehicleSAInterfaceVTBL*)GetVehicleInterface()->vtbl)->GetDoorAngleOpenRatio_;
-    if (dwFunc == FUNC_CAutomobile__GetDoorAngleOpenRatio)
-        dwOffset = 1464;
-    else if (dwFunc == FUNC_CTrain__GetDoorAngleOpenRatio)
-        dwOffset = 1496;
-    else
-        dwOffset = 0;            // Other vehicles don't have door information.
+    // Initialize doors depending on the vehicle type.
+    DWORD dwOffset = 0;
+
+    switch (static_cast<VehicleClass>(GetVehicleInterface()->m_vehicleClass))
+    {
+        case VehicleClass::AUTOMOBILE:
+        case VehicleClass::MONSTER_TRUCK:
+        case VehicleClass::QUAD:
+        case VehicleClass::HELI:
+        case VehicleClass::PLANE:
+        case VehicleClass::TRAILER:
+        {
+            dwOffset = 1464;
+        }
+        case VehicleClass::TRAIN:
+        {
+            dwOffset = 1496;
+        }
+        default:
+            break;
+    }
 
     if (dwOffset != 0)
     {
         for (unsigned int i = 0; i < sizeof(m_doors) / sizeof(m_doors[0]); ++i)
         {
             DWORD dwInterface = (DWORD)GetInterface();
-            DWORD dwDoorAddress = dwInterface + 1464 + i * 24;
+            DWORD dwDoorAddress = dwInterface + dwOffset + i * 24;
             m_doors[i].SetInterface((CDoorSAInterface*)dwDoorAddress);
         }
     }
```

Comment:
This code was bugged. The game uses wrong door array positions for CTrain.
It causes train desync issues